In [30]:
%run '00_database_connectivity_setup.ipynb'

x         y
0  0.778415  0.134201
1  0.902056  0.791205
2  0.751707  0.972293
3  0.788108  0.152631
4  0.306569  0.419736

In [31]:
%%execsql
set search_path to yelp;


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import json
from IPython.display import display

%matplotlib inline

# Limit rows disp# Limit rows displayed in notebook
pd.set_option('display.max_rows', 50)
pd.set_option('display.max_columns', 300)
pd.set_option('display.precision', 6)

/Users/asong/anaconda/lib/python2.7/site-packages/pandas/computation/__init__.py:19: UserWarning: The installed version of numexpr 2.4.4 is not supported in pandas and will be not be used

  UserWarning)


## Notes on Data

__The Challenge Dataset:__

+ 2.2M reviews and 591K tips by 552K users for 77K businesses
+ 566K business attributes, e.g., hours, parking availability, ambience.
+ Social network of 552K users for a total of 3.5M social edges.
+ Aggregated check-ins over time for each of the 77K businesses
+ 200,000 pictures from the included businesses

__Cities:__

+ U.K.: Edinburgh
+ Germany: Karlsruhe
+ Canada: Montreal and Waterloo
+ U.S.: Pittsburgh, Charlotte, Urbana-Champaign, Phoenix, Las Vegas, Madison


### business
```json
{
    'type': 'business',
    'business_id': (encrypted business id),
    'name': (business name),
    'neighborhoods': [(hood names)],
    'full_address': (localized address),
    'city': (city),
    'state': (state),
    'latitude': latitude,
    'longitude': longitude,
    'stars': (star rating, rounded to half-stars),
    'review_count': review count,
    'categories': [(localized category names)]
    'open': True / False (corresponds to closed, not business hours),
    'hours': {
        (day_of_week): {
            'open': (HH:MM),
            'close': (HH:MM)
        },
        ...
    },
    'attributes': {
        (attribute_name): (attribute_value),
        ...
    },
}
```

### review
```json
{
    'type': 'review',
    'business_id': (encrypted business id),
    'user_id': (encrypted user id),
    'stars': (star rating, rounded to half-stars),
    'text': (review text),
    'date': (date, formatted like '2012-03-14'),
    'votes': {(vote type): (count)},
    'review_id': (encrypted review id)
}
```

### user
```json
{
    'type': 'user',
    'user_id': (encrypted user id),
    'name': (first name),
    'review_count': (review count),
    'average_stars': (floating point average, like 4.31),
    'votes': {(vote type): (count)},
    'friends': [(friend user_ids)],
    'elite': [(years_elite)],
    'yelping_since': (date, formatted like '2012-03'),
    'compliments': {
        (compliment_type): (num_compliments_of_this_type),
        ...
    },
    'fans': (num_fans),
}
```

### check-in
```json
{
    'type': 'checkin',
    'business_id': (encrypted business id),
    'checkin_info': {
        '0-0': (number of checkins from 00:00 to 01:00 on all Sundays),
        '1-0': (number of checkins from 01:00 to 02:00 on all Sundays),
        ...
        '14-4': (number of checkins from 14:00 to 15:00 on all Thursdays),
        ...
        '23-6': (number of checkins from 23:00 to 00:00 on all Saturdays)
    }, # if there was no checkin for a hour-day block it will not be in the dict
}
```

### tip
```json
{
    'type': 'tip',
    'text': (tip text),
    'business_id': (encrypted business id),
    'user_id': (encrypted user id),
    'date': (date, formatted like '2012-03-14'),
    'likes': (count),
}
```

## Business Data

In [194]:
%%showsql
select * from yelp.business_raw limit 10

attributes  \
0  {"By Appointment Only": false, "Parking": {"garage": false, "street": false, "validated": false, "lot": false, "valet": false}, "Wheelchair Accessible": true}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
1  {"Accepts Credit Cards": true, "Price Range": 2, "Parking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
2  {"Take-out": true, "Drive-Thru": false, "Outdoor Seating": false, "Caters": false, "Noise Level": "average", "Parking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}, "Good For Groups": true, "Delivery": false, "Attire": "casual", "Has TV": true, "Price Range": 1, "Good For": {"dessert": false, "latenight": false, "lunch": false, "dinner": false, "brunch": false, "breakfast": true}, "Takes Reservations": false, "Ambience": {"romantic": false, "intimate": false, "classy": false, "hipster": false, "divey": false, "touristy": false, "trendy": false, "upscale": false, "casual": true}, "Waiter Service": true, "Accepts Credit Cards": true, "Good for Kids": true, "Alcohol": "none", "Wheelchair Accessible": true}                    
3  {"Price Range": 3, "Accepts Credit Cards": true, "Wheelchair Accessible": true, "Parking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
4  {"Accepts Credit Cards": true}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
5  {"Take-out": true, "Accepts Credit Cards": true, "Price Range": 1, "Outdoor Seating": false, "Caters": 

### Categories

In [28]:
%%execsql
create or replace function split_array(
    arr text
) returns setof text as
$$
    try:
        arr_split = arr[1:-1].replace('"','').split(',')
        if not arr_split[0]:
            return []
        else:
            return tuple(arr_split)
    except:
        return []
$$ language plpythonu;

create or replace function split_array(
    arr text[]
) returns setof text as
$$
    return tuple(arr)
$$ language plpythonu;

In [26]:
%%execsql
drop table if exists business_categories;
create table business_categories as
select
    trim(both '"' from business_id) as business_id
    ,trim(both ' ' from split_array(categories)) as category
from business_raw 
distributed by (business_id);

In [12]:
%%showsql
select * from business_categories limit 5;

business_id      category
0  TLrX4FECQDFz-E45TI_9ow  Seafood     
1  TLrX4FECQDFz-E45TI_9ow   Restaurants
2  WUtPI9rJbs7ET1gPhULnDw  Southern    
3  WUtPI9rJbs7ET1gPhULnDw   Restaurants
4  iLwNx5TVKVFKyV5RdbIY-A  Sushi Bars

## Limit scope of data to restaurants in US and Canada

In [36]:
%%execsql
drop table if exists restaurants_us_ca;
create table restaurants_us_ca as
select 
    *
    ,case when state='AZ' then 'Phoenix'
    when state='PA' then 'Pittsburgh'
    when state='IL' then 'Urbana-Champaign'
    when state='NV' then 'Las Vegas'
    when state='WI' then 'Madison'
    when state='NC' then 'Charlotte'
    when state='ON' then 'Waterloo'
    when state='QC' then 'Montreal'
    else null end as city_1
from 
    business_raw
join (
    select 
        '"'||business_id||'"' as business_id 
    from 
        business_categories 
    where 
        category ilike ANY(array['restaurants', 'food']) 
    group by 1
) cats
using (business_id)
where 
    state ilike ANY(array['AZ','PA','IL','NV','WI', 'NC', 'ON','QC'])
distributed by (business_id);

In [43]:
%%showsql
select 
    *
    ,case when state='AZ' then 'Phoenix'
    when state='PA' then 'Pittsburgh'
    when state='IL' then 'Urbana-Champaign'
    when state='NV' then 'Las Vegas'
    when state='WI' then 'Madison'
    when state='NC' then 'Charlotte'
    when state='ON' then 'Waterloo'
    when state='QC' then 'Montreal'
    else null end as city_1
from 
    business_raw
join (
    select 
        '"'||business_id||'"' as business_id 
    from 
        business_categories 
    where 
        category ilike ANY(array['restaurants', 'food']) 
    group by 1
) cats
using (business_id) limit 5

business_id  \
0  "2PfavOTufsPCRdYm-bFcpw"   
1  "2VuPoaFPktvtAqTX_Fzc-A"   
2  "lX3XYcMWoRlHjrmkoGNefQ"   
3  "bewtZL-VV1oeGUdAhHVz6Q"   
4  "sAbRiVqmxNdjTja_vmnsnA"   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          attributes  \
0  {"Price Range": 1, "Accepts Credit Cards": false, "Wheelchair Accessible": true, "Wi-Fi": "no", "Parking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}}                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
1  {"Take-out": true, "Accepts Credit Cards": true, "Price Range": 1, "Outdoor Seating": false, "Caters": false, "Noise Level": "average", "Parking": {"garage": false, "street": false, "validated": false, "lot": false, "valet": false}, "Delivery": false, "Attire": "casual", "Has TV": true, "Good For": {"dessert": false, "latenight": false, "lunch": false, "dinner": false, "brunch": false, "breakfast": false}, "Takes Reservations": false, "Ambience": {"romantic": false, "intimate": false, "touristy": false, "hipster": false, "divey": false, "classy": false, "trendy": false, "upscale": false, "casual": false}, "Waiter Service": true, "Wi-Fi": "no", "Good for Kids": true, "Good For Groups": true, "Alcohol": "full_bar"}                                                                                                                  
2  {"Take-out": true, "Accepts Credit Cards": true, "Price Range": 1, "Outdoor Seating": false, "Caters": true, "Noise Level": "average", "Parking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}, "Good For Groups": true, "Delivery": true, "Dogs Allowed": false, "Attire": "casual", "Has TV": true, "Good For": {"dessert": false, "latenight": false, "lunch": true, "dinner": false, "brunch": false, "breakfast": false}, "Takes Reservations": false, "Ambience": {"romantic": false, "intimate": false, "classy": false, "hipster": false, "divey": false, "touristy": false, "trendy": false, "upscale": false, "casual": true}, "Waiter Service": false, "Wi-Fi": "free", "Good for Kids": true, "Alcohol": "none", "Wheelchair Accessible": false}                                                                 
3  {"Take-out": false, "Price Range": 3, "Outdoor Seating": false, "Noise Level": "quiet", "Parking": {"garage": false, "street": false, "validated": false, "lot": true, "valet": false}, "Delivery": false, "Attire": "casual", "Has TV": false, "Good For": {"dessert": false, "latenight": false, "lunch": false, "dinner": false, "breakfast": false, "brunch": false}, "Takes Reservations": true, "Ambience": {"romantic": false, "intimate": false, "touristy": false, "hipst

### attributes
Attributes are json objects. Parse these into a separate table

In [39]:
%%execsql
drop type if exists attributes cascade;
create type attributes as (
    key text
    ,value text
);

create or replace function flatten_json(
    json_str text
    ,prefix text
) returns setof attributes as 
$$
    import json
    
    def flatten_json(json_dict, data =[], key=''):    
        for i in json_dict.items():
            if type(i[1]) != dict:
                data.append([(key+'.'+i[0]).replace(' ','_'), i[1]])
            else:
                data.extend(flatten_json(i[1], [], key+'.'+i[0]))
        return data

    return tuple(flatten_json(json.loads(json_str), key=prefix))
    
$$ language plpythonu;

In [99]:
%%execsql
-- table of attributes unnested 
drop table if exists business_attributes_unnest;
create table business_attributes_unnest as
with attributes as ( -- table of attributes unnested and any duplicates removed
    select 
        trim(both '"' from business_id) as business_id
        ,lower((attr).key) as key
        ,lower((attr).value) as value
    from (
        select
            business_id
            ,flatten_json(attributes, 'attr') as attr
        from business_raw
    ) t1
    GROUP BY 1,2,3
)
select 
     business_id
    ,key
    ,value
from 
    attributes
where key !='attr.good_for_kids'
UNION 
select 
    business_id
    ,key
    ,array_to_string(array_agg(value),',') AS value
from 
    attributes
where key ='attr.good_for_kids'
group by 1,2
distributed by (key);

In [100]:
%%showsql
select * from business_attributes_unnest limit 5;

business_id               key  value
0  1wIKHSHPOkywuOkyrXt6dw  attr.music.video  false
1  5BXNeSwzOy1nhU9PtCHvIw  attr.music.video  false
2  cLojUtqkFpVIHFEvA_5Atw  attr.music.video  false
3  fytNHpJ5Js9hkD_cQO8rBA  attr.music.video  false
4  gTm2TSYkC44iMnI_Nzpm4Q  attr.music.video  false

__Let's take a look at the different attributes available__

In [139]:
%%showsql
select
    key
    ,count(*)
from 
    business_attributes_unnest
group by 
    1
order by 1;

key  count
0   attr.accepts_credit_cards                       45067
1   attr.accepts_insurance                          458  
2   attr.ages_allowed                               235  
3   attr.alcohol                                    20457
4   attr.ambience.casual                            18531
5   attr.ambience.classy                            18531
6   attr.ambience.divey                             17546
7   attr.ambience.hipster                           18309
8   attr.ambience.intimate                          18531
9   attr.ambience.romantic                          18531
10  attr.ambience.touristy                          18531
11  attr.ambience.trendy                            18531
12  attr.ambience.upscale                           18396
13  attr.attire                                     20390
14  attr.by_appointment_only                        7922 
15  attr.byob                                       866  
16  attr.byob/corkage                               1315 
17  attr.caters                                     13116
18  attr.coat_check                                 3699 
19  attr.corkage                                    655  
20  attr.delivery                                   19924
21  attr.dietary_restrictions.dairy-free            175  
22  attr.dietary_restrictions.gluten-free           175  
23  attr.dietary_restrictions.halal                 175  
24  attr.dietary_restrictions.kosher                175  
25  attr.dietary_restrictions.soy-free              175  
26  attr.dietary_restrictions.vegan                 175  
27  attr.dietary_restrictions.vegetarian            175  
28  attr.dogs_allowed                               3591 
29  attr.drive-thru                                 2569 
30  attr.good_for.breakfast                         19267
31  attr.good_for.brunch                            19204
32  attr.good_for_dancing                           3581 
33  attr.good_for.dessert                           19196
34  attr.good_for.dinner                            19257
35  attr.good_for_groups                            22406
36  attr.good_for_kids                              25525
37  attr.good_for.latenight                         19258
38  attr.good_for.lunch                             19257
39  attr.hair_types_specialized_in.africanamerican  318  
40  attr.hair_types_specialized_in.asian            318  
41  attr.hair_types_specialized_in.coloring         326  
42  attr.hair_types_specialized_in.curly            326  
43  attr.hair_types_specialized_in.extensions       326  
44  attr.hair_types_specialized_in.kids             326  
45  attr.hair_types_specialized_in.perms            326  
46  attr.hair_types_specialized_in.straightperms    318  
47  attr.happy_hour                                 3728 
48  attr.has_tv                                     19338
49  attr.music.background_music                     1586 
50  attr.music.dj                                   3278 
51  attr.music.jukebox                              2724 
52  attr.music.karaoke                              1598 
53  attr.music.live                                 2720 
54  attr.music.playlist                             2    
55  attr.music.video                                2584 
56  attr.noise_level                                18519
57  attr.open_24_hours                              334  
58  attr.order_at_counter                           384  
59  attr.outdoor_seating                            23033
60  attr.parking.garage                             37944
61  attr.parking.lot                                37926
62  attr.parking.street                             37926
63  attr.parking.valet                              37929
64  attr.parking.validated                          37523
65  attr.payment_types.amex                         13   
66  attr.payment_types.cash_only                    13   
67  attr.payment_types.discover                     13   
68  attr.payment_types.mastercard            

__good_for attributes__

Looks like there are two types of "good for" attributes: 

1) good for type of eating
    + attr.good_for.breakfast
    + attr.good_for.brunch
    + attr.good_for.dessert
    + attr.good_for.dinner
    + attr.good_for.latenight
    + attr.good_for.lunch
2) good for abience/activity
    + attr.good_for_dancing
    + attr.good_for_groups
    + attr.good_for_kids

__Pivot the attributes table__

In [6]:
%%execsql
-- convert booleans to binary (in text for now)
drop table if exists business_attributes_unnest_temp;
create temp table business_attributes_unnest_temp as
select
    business_id
    ,key
    ,case when value='false' then '0'
     when value='true' then '1'
    else value end as value
from 
    business_attributes_unnest
distributed by (business_id);


-- this table is going to have boolean values (0,1) as text
drop table if exists business_attributes_temp;
select sugarlib.sugar($$
    create temp table business_attributes_temp as
    select business_id
        ,sugarlib.pivot(key,business_attributes_unnest_temp,value)
    from 
        business_attributes_unnest_temp
    group by business_id
    distributed by (business_id);
$$);

In [12]:
%%showsql
select * from business_attributes_temp limit 5;

business_id key_attr.accepts_credit_cards  \
0  08xNB6lB-2kMrj5bqJaEMA  1                              
1  0Ayzd0haow5Ek3CdfZg5bA  1                              
2  0btslkqA-YJQjLH6GFjU-g  1                              
3  0gABc2O0CsMqhT35LPHfNQ  1                              
4  0m-dEwpqVNO4mgZ1mM7Wew  1                              

  key_attr.accepts_insurance key_attr.ages_allowed key_attr.alcohol  \
0  None                       None                  None              
1  None                       None                  beer_and_wine     
2  None                       None                  None              
3  None                       None                  None              
4  None                       None                  None              

  key_attr.ambience.casual key_attr.ambience.classy key_attr.ambience.divey  \
0  None                     None                     None                     
1  0                        0                        0                        
2  None                     None                     None                     
3  None                     None                     None                     
4  None                     None                     None                     

  key_attr.ambience.hipster key_attr.ambience.intimate  \
0  None                      None                        
1  0                         0                           
2  None                      None                        
3  None                      None                        
4  None                      None                        

  key_attr.ambience.romantic key_attr.ambience.touristy  \
0  None                       None                        
1  0                          1                           
2  None                       None                        
3  None                       None                        
4  None                       None                        

  key_attr.ambience.trendy key_attr.ambience.upscale key_attr.attire  \
0  None                     None                      None             
1  0                        0                         casual           
2  None                     None                      None             
3  None                     None                      None             
4  None                     None                      None             

  key_attr.by_appointment_only key_attr.byob key_attr.byob/corkage  \
0  None                         None          None                   
1  None                         None          None                   
2  None                         None          None                   
3  None                         None          None                   
4  None                         None          None                   

  key_attr.caters key_attr.coat_check key_attr.corkage key_attr.delivery  \
0  None            None                None             None               
1  0               None                None             0                  
2  None            None                None             None               
3  None            None                None             None               
4  None            None                None             None               

  key_attr.dietary_restrictions.dairy-free  \
0  None                                      
1  None                                      
2  None                                      
3  None                                      
4  None                                      

  key_attr.dietary_restrictions.gluten-free  \
0  None                                       
1  None                                       
2  None                                       
3  None                                       
4  None                                       

  key_attr.dietary_restrictions.halal key_attr.dietary_restrictions.kosher  \
0  None                                None                                  
1  None                   

In [13]:
%%showsql
-- binary boolean values that need to be casted to integer type
select key from business_attributes_unnest where value not in ('true', 'false') group by 1;

key
0  attr.byob/corkage 
1  attr.smoking      
2  attr.wi-fi        
3  attr.ages_allowed 
4  attr.alcohol      
5  attr.good_for_kids
6  attr.noise_level  
7  attr.price_range  
8  attr.attire

In [16]:
%%execsql
drop table if exists business_attributes;
create table business_attributes as 
select
    business_id
    ,"key_attr.payment_types.visa"::integer as "payment_types.visa"
    ,"key_attr.hair_types_specialized_in.kids"::integer as "hair_types_specialized_in.kids"
    ,"key_attr.dietary_restrictions.dairy-free"::integer as "dietary_restrictions.dairy-free"
    ,"key_attr.hair_types_specialized_in.perms"::integer as "hair_types_specialized_in.perms"
    ,"key_attr.dietary_restrictions.kosher"::integer as "dietary_restrictions.kosher"
    ,"key_attr.hair_types_specialized_in.straightperms"::integer as "hair_types_specialized_in.straightperms"
    ,"key_attr.music.karaoke"::integer as "music.karaoke"
    ,"key_attr.payment_types.discover"::integer as "payment_types.discover"
    ,"key_attr.order_at_counter"::integer as "order_at_counter"
    ,"key_attr.dietary_restrictions.halal"::integer as "dietary_restrictions.halal"
    ,"key_attr.dietary_restrictions.gluten-free"::integer as "dietary_restrictions.gluten-free"
    ,"key_attr.payment_types.amex"::integer as "payment_types.amex"
    ,"key_attr.happy_hour"::integer as "happy_hour"
    ,"key_attr.hair_types_specialized_in.extensions"::integer as "hair_types_specialized_in.extensions"
    ,"key_attr.corkage"::integer as "corkage"
    ,"key_attr.dietary_restrictions.vegetarian"::integer as "dietary_restrictions.vegetarian"
    ,"key_attr.hair_types_specialized_in.curly"::integer as "hair_types_specialized_in.curly"
    ,"key_attr.accepts_insurance"::integer as "accepts_insurance"
    ,"key_attr.music.playlist"::integer as "music.playlist"
    ,"key_attr.good_for_dancing"::integer as "good_for_dancing"
    ,"key_attr.byob"::integer as "byob"
    ,"key_attr.hair_types_specialized_in.asian"::integer as "hair_types_specialized_in.asian"
    ,"key_attr.music.video"::integer as "music.video"
    ,"key_attr.payment_types.cash_only"::integer as "payment_types.cash_only"
    ,"key_attr.hair_types_specialized_in.africanamerican"::integer as "hair_types_specialized_in.africanamerican"
    ,"key_attr.by_appointment_only"::integer as "by_appointment_only"
    ,"key_attr.caters"::integer as "caters"
    ,"key_attr.ambience.intimate"::integer as "ambience.intimate"
    ,"key_attr.ambience.upscale"::integer as "ambience.upscale"
    ,"key_attr.ambience.divey"::integer as "ambience.divey"
    ,"key_attr.waiter_service"::integer as "waiter_service"
    ,"key_attr.open_24_hours"::integer as "open_24_hours"
    ,"key_attr.ambience.trendy"::integer as "ambience.trendy"
    ,"key_attr.payment_types.mastercard"::integer as "payment_types.mastercard"
    ,"key_attr.good_for.breakfast"::integer as "good_for.breakfast"
    ,"key_attr.ambience.classy"::integer as "ambience.classy"
    ,"key_attr.ambience.romantic"::integer as "ambience.romantic"
    ,"key_attr.ambience.touristy"::integer as "ambience.touristy"
    ,"key_attr.good_for.lunch"::integer as "good_for.lunch"
    ,"key_attr.outdoor_seating"::integer as "outdoor_seating"
    ,"key_attr.wheelchair_accessible"::integer as "wheelchair_accessible"
    ,"key_attr.music.jukebox"::integer as "music.jukebox"
    ,"key_attr.music.background_music"::integer as "music.background_music"
    ,"key_attr.has_tv"::integer as "has_tv"
    ,"key_attr.take-out"::integer as "take-out"
    ,"key_attr.drive-thru"::integer as "drive-thru"
    ,"key_attr.good_for_groups"::integer as "good_for_groups"
    ,"key_attr.dietary_restrictions.soy-free"::integer as "dietary_restrictions.soy-free"
    ,"key_attr.takes_reservations"::integer as "takes_reservations"
    ,"key_attr.delivery"::integer as "delivery"
    ,"key_attr.coat_check"::integer as "coat_check"
    ,"key_attr.parking.valet"::integer as "parking.valet"
    ,"key_attr.parking.lot"::integer as "parking.lot"
    ,"key_attr.dietary_restrictions.vegan"::integer as "dietary_restrictions.vegan"
    ,"key_attr.good_for.brunch"::integer as "good_for.brunch"
    ,"key_attr.good_for.dinner"::integer as "good_for.dinner"
    ,"key_attr.hair_types_specialized_in.coloring"::integer as "hair_types_specialized_in.coloring"
    ,"key_attr.music.dj"::integer as "music.dj"
    ,"key_attr.good_for.dessert"::integer as "good_for.dessert"
    ,"key_attr.good_for.latenight"::integer as "good_for.latenight"
    ,"key_attr.parking.garage"::integer as "parking.garage"
    ,"key_attr.dogs_allowed"::integer as "dogs_allowed"
    ,"key_attr.parking.validated"::integer as "parking.validated"
    ,"key_attr.music.live"::integer as "music.live"
    ,"key_attr.parking.street"::integer as "parking.street"
    ,case when "key_attr.good_for_kids"='1' then 1 
    when "key_attr.good_for_kids"='0' then 0
    else 2 end as "good_for_kids" -- 2 means it was both true and false in the json ie ambiguous
    ,"key_attr.ambience.hipster"::integer as "ambience.hipster"
    ,"key_attr.accepts_credit_cards"::integer as "accepts_credit_cards"
    ,"key_attr.ambience.casual"::integer as "ambience.casual"
    ,"key_attr.byob/corkage" as "byob/corkage"
    ,"key_attr.smoking" as "smoking"
    ,"key_attr.wi-fi" as "wi-fi"
    ,"key_attr.ages_allowed" as "ages_allowed"
    ,"key_attr.alcohol" as "alcohol"
    ,"key_attr.noise_level" as "noise_level"
    ,"key_attr.price_range" as "price_range"
    ,"key_attr.attire" as "attire"
from 
    business_attributes_temp
distributed by (business_id);

In [200]:
%%execsql
drop type if exists count_type cascade;
create type count_type as (
    col text
    ,cnt integer
    ,prob double precision
    ,cume_prob double precision
);

create or replace function count_summary(
    col text
    ,table_name text
) returns setof count_type as
$$
    import ast
    sql = """
        with summary as (
            select 
                '"'||"{col}"||'"' as col
                ,count(*) as cnt
                ,count(*)/sum(count(*)) OVER () as prob
                ,sum(count(*)) over(order by count(*) desc rows unbounded preceding) / sum(count(*)) over () as cume_prob
            from 
                {table_name}
            group by 1
        )
        select summary from summary;
    """.format(col=col, table_name=table_name)
    res = plpy.execute(sql)
    
    results = []
    for rec in res:
        s = rec['summary'].replace('(,', '(None,')
        plpy.info(s)
        results.append(ast.literal_eval(s))
    return tuple(results)
$$ language plpythonu;

In [219]:
sql = """select column_name::text from information_schema.columns where table_name='business_attributes' and table_schema='yelp' order by ordinal_position"""

df = psql.read_sql(sql,conn)
conn.commit()
columns = df.column_name.tolist()[1:]
table_name = 'yelp.business_attributes'
fails = []

for col in columns:
    try:
        sql="""select * from count_summary('{col}', '{table_name}');""".format(table_name=table_name, col=col)
        df = psql.read_sql(sql,conn)
        conn.commit()
        display(col)
        display(df)
    except:
        fails.append(col)
        pass

'key_attr.accepts_credit_cards'

col    cnt      prob  cume_prob
0  1     42889  0.814745  0.814745 
1  None  7574   0.143880  0.958625 
2  0     2178   0.041375  1.000000

'key_attr.accepts_insurance'

col    cnt      prob  cume_prob
0  None  52183  0.991300  0.991300 
1  0     391    0.007428  0.998727 
2  1     67     0.001273  1.000000

'key_attr.ages_allowed'

col    cnt      prob  cume_prob
0  None     52406  0.995536  0.995536 
1  21plus   168    0.003191  0.998727 
2  18plus   43     0.000817  0.999544 
3  allages  19     0.000361  0.999905 
4  19plus   5      0.000095  1.000000

'key_attr.alcohol'

col    cnt      prob  cume_prob
0  None           32184  0.611387  0.611387 
1  full_bar       9069   0.172280  0.783667 
2  none           8405   0.159666  0.943333 
3  beer_and_wine  2983   0.056667  1.000000

'key_attr.ambience.casual'

col    cnt      prob  cume_prob
0  None  34110  0.647974  0.647974 
1  0     10193  0.193632  0.841606 
2  1     8338   0.158394  1.000000

'key_attr.ambience.classy'

col    cnt      prob  cume_prob
0  None  34110  0.647974  0.647974 
1  0     18092  0.343686  0.991660 
2  1     439    0.008340  1.000000

'key_attr.ambience.divey'

col    cnt      prob  cume_prob
0  None  35095  0.666686  0.666686 
1  0     16685  0.316958  0.983644 
2  1     861    0.016356  1.000000

'key_attr.ambience.hipster'

col    cnt      prob  cume_prob
0  None  34332  0.652191  0.652191 
1  0     17965  0.341274  0.993465 
2  1     344    0.006535  1.000000

'key_attr.ambience.intimate'

col    cnt      prob  cume_prob
0  None  34110  0.647974  0.647974 
1  0     18316  0.347942  0.995916 
2  1     215    0.004084  1.000000

'key_attr.ambience.romantic'

col    cnt      prob  cume_prob
0  None  34110  0.647974  0.647974 
1  0     18281  0.347277  0.995251 
2  1     250    0.004749  1.000000

'key_attr.ambience.touristy'

col    cnt      prob  cume_prob
0  None  34110  0.647974  0.647974 
1  0     18399  0.349518  0.997492 
2  1     132    0.002508  1.000000

'key_attr.ambience.trendy'

col    cnt      prob  cume_prob
0  None  34110  0.647974  0.647974 
1  0     17810  0.338329  0.986303 
2  1     721    0.013697  1.000000

'key_attr.ambience.upscale'

col    cnt      prob  cume_prob
0  None  34245  0.650539  0.650539 
1  0     18228  0.346270  0.996809 
2  1     168    0.003191  1.000000

'key_attr.attire'

col    cnt      prob  cume_prob
0  None    32251  0.612659  0.612659 
1  casual  19551  0.371403  0.984062 
2  dressy  775    0.014722  0.998784 
3  formal  64     0.001216  1.000000

'key_attr.by_appointment_only'

col    cnt      prob  cume_prob
0  None  44719  0.849509  0.849509 
1  0     4482   0.085143  0.934652 
2  1     3440   0.065348  1.000000

'key_attr.byob'

col    cnt      prob  cume_prob
0  None  51775  0.983549  0.983549 
1  0     821    0.015596  0.999145 
2  1     45     0.000855  1.000000

'key_attr.byob/corkage'

col    cnt      prob  cume_prob
0  None         51326  0.975019  0.975019 
1  no           693    0.013165  0.988184 
2  yes_free     476    0.009042  0.997226 
3  yes_corkage  146    0.002774  1.000000

'key_attr.caters'

col    cnt      prob  cume_prob
0  None  39525  0.750841  0.750841 
1  0     6869   0.130488  0.881328 
2  1     6247   0.118672  1.000000

'key_attr.coat_check'

col    cnt      prob  cume_prob
0  None  48942  0.929732  0.929732 
1  0     3292   0.062537  0.992268 
2  1     407    0.007732  1.000000

'key_attr.corkage'

col    cnt      prob  cume_prob
0  None  51986  0.987557  0.987557 
1  0     518    0.009840  0.997397 
2  1     137    0.002603  1.000000

'key_attr.delivery'

col    cnt      prob  cume_prob
0  None  32717  0.621512  0.621512 
1  0     16175  0.307270  0.928782 
2  1     3749   0.071218  1.000000

'key_attr.dietary_restrictions.dairy-free'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  0     158    0.003001  0.999677 
2  1     17     0.000323  1.000000

'key_attr.dietary_restrictions.gluten-free'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  0     166    0.003153  0.999829 
2  1     9      0.000171  1.000000

'key_attr.dietary_restrictions.halal'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  0     170    0.003229  0.999905 
2  1     5      0.000095  1.000000

'key_attr.dietary_restrictions.kosher'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  0     171    0.003248  0.999924 
2  1     4      0.000076  1.000000

'key_attr.dietary_restrictions.soy-free'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  0     169    0.003210  0.999886 
2  1     6      0.000114  1.000000

'key_attr.dietary_restrictions.vegan'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  1     90     0.001710  0.998385 
2  0     85     0.001615  1.000000

'key_attr.dietary_restrictions.vegetarian'

col    cnt      prob  cume_prob
0  None  52466  0.996676  0.996676 
1  0     123    0.002337  0.999012 
2  1     52     0.000988  1.000000

'key_attr.dogs_allowed'

col    cnt      prob  cume_prob
0  None  49050  0.931783  0.931783 
1  0     2426   0.046086  0.977869 
2  1     1165   0.022131  1.000000

'key_attr.drive-thru'

col    cnt      prob  cume_prob
0  None  50072  0.951198  0.951198 
1  0     1382   0.026253  0.977451 
2  1     1187   0.022549  1.000000

'key_attr.good_for.breakfast'

col    cnt      prob  cume_prob
0  None  33374  0.633993  0.633993 
1  0     17743  0.337057  0.971049 
2  1     1524   0.028951  1.000000

'key_attr.good_for.brunch'

col    cnt      prob  cume_prob
0  None  33437  0.635189  0.635189 
1  0     18243  0.346555  0.981744 
2  1     961    0.018256  1.000000

'key_attr.good_for_dancing'

col    cnt      prob  cume_prob
0  None  49060  0.931973  0.931973 
1  0     2621   0.049790  0.981763 
2  1     960    0.018237  1.000000

'key_attr.good_for.dessert'

col    cnt      prob  cume_prob
0  None  33445  0.635341  0.635341 
1  0     18875  0.358561  0.993902 
2  1     321    0.006098  1.000000

'key_attr.good_for.dinner'

col    cnt      prob  cume_prob
0  None  33384  0.634182  0.634182 
1  0     13689  0.260044  0.894227 
2  1     5568   0.105773  1.000000

'key_attr.good_for_groups'

col    cnt      prob  cume_prob
0  None  30235  0.574362  0.574362 
1  1     19898  0.377994  0.952357 
2  0     2508   0.047643  1.000000

'key_attr.good_for_kids'

col    cnt      prob  cume_prob
0  None        27116  0.515112  0.515112 
1  1           19938  0.378754  0.893866 
2  0           5297   0.100625  0.994491 
3  true,false  146    0.002774  0.997264 
4  false,true  144    0.002736  1.000000

'key_attr.good_for.latenight'

col    cnt      prob  cume_prob
0  None  33383  0.634163  0.634163 
1  0     18257  0.346821  0.980984 
2  1     1001   0.019016  1.000000

'key_attr.good_for.lunch'

col    cnt      prob  cume_prob
0  None  33384  0.634182  0.634182 
1  0     12085  0.229574  0.863756 
2  1     7172   0.136244  1.000000

'key_attr.hair_types_specialized_in.africanamerican'

col    cnt      prob  cume_prob
0  None  52323  0.993959  0.993959 
1  0     204    0.003875  0.997834 
2  1     114    0.002166  1.000000

'key_attr.hair_types_specialized_in.asian'

col    cnt      prob  cume_prob
0  None  52323  0.993959  0.993959 
1  1     167    0.003172  0.997132 
2  0     151    0.002868  1.000000

'key_attr.hair_types_specialized_in.coloring'

col    cnt      prob  cume_prob
0  None  52315  0.993807  0.993807 
1  1     292    0.005547  0.999354 
2  0     34     0.000646  1.000000

'key_attr.hair_types_specialized_in.curly'

col    cnt      prob  cume_prob
0  None  52315  0.993807  0.993807 
1  1     246    0.004673  0.998480 
2  0     80     0.001520  1.000000

'key_attr.hair_types_specialized_in.extensions'

col    cnt      prob  cume_prob
0  None  52315  0.993807  0.993807 
1  1     198    0.003761  0.997568 
2  0     128    0.002432  1.000000

'key_attr.hair_types_specialized_in.kids'

col    cnt      prob  cume_prob
0  None  52315  0.993807  0.993807 
1  1     163    0.003096  0.996904 
2  0     163    0.003096  1.000000

'key_attr.hair_types_specialized_in.perms'

col    cnt      prob  cume_prob
0  None  52315  0.993807  0.993807 
1  0     176    0.003343  0.997151 
2  1     150    0.002849  1.000000

'key_attr.hair_types_specialized_in.straightperms'

col    cnt      prob  cume_prob
0  None  52323  0.993959  0.993959 
1  0     179    0.003400  0.997359 
2  1     139    0.002641  1.000000

'key_attr.happy_hour'

col    cnt      prob  cume_prob
0  None  48913  0.929181  0.929181 
1  1     2640   0.050151  0.979332 
2  0     1088   0.020668  1.000000

'key_attr.has_tv'

col    cnt      prob  cume_prob
0  None  33303  0.632644  0.632644 
1  1     9955   0.189111  0.821755 
2  0     9383   0.178245  1.000000

'key_attr.music.background_music'

col    cnt      prob  cume_prob
0  None  51055  0.969871  0.969871 
1  1     803    0.015254  0.985126 
2  0     783    0.014874  1.000000

'key_attr.music.dj'

col    cnt      prob  cume_prob
0  None  49363  0.937729  0.937729 
1  0     2308   0.043844  0.981573 
2  1     970    0.018427  1.000000

'key_attr.music.jukebox'

col    cnt      prob  cume_prob
0  None  49917  0.948253  0.948253 
1  0     2026   0.038487  0.986740 
2  1     698    0.013260  1.000000

'key_attr.music.karaoke'

col    cnt      prob  cume_prob
0  None  51043  0.969643  0.969643 
1  0     1500   0.028495  0.998138 
2  1     98     0.001862  1.000000

'key_attr.music.live'

col    cnt      prob  cume_prob
0  None  49921  0.948329  0.948329 
1  0     1907   0.036227  0.984556 
2  1     813    0.015444  1.000000

'key_attr.music.playlist'

col    cnt      prob  cume_prob
0  None  52639  0.999962  0.999962 
1  1     1      0.000019  0.999981 
2  0     1      0.000019  1.000000

'key_attr.music.video'

col    cnt      prob  cume_prob
0  None  50057  0.950913  0.950913 
1  0     2421   0.045991  0.996904 
2  1     163    0.003096  1.000000

'key_attr.noise_level'

col    cnt      prob  cume_prob
0  None       34122  0.648202  0.648202 
1  average    11754  0.223286  0.871488 
2  quiet      3996   0.075910  0.947398 
3  loud       1818   0.034536  0.981934 
4  very_loud  951    0.018066  1.000000

'key_attr.open_24_hours'

col    cnt      prob  cume_prob
0  None  52307  0.993655  0.993655 
1  0     300    0.005699  0.999354 
2  1     34     0.000646  1.000000

'key_attr.order_at_counter'

col    cnt      prob  cume_prob
0  None  52257  0.992705  0.992705 
1  1     241    0.004578  0.997283 
2  0     143    0.002717  1.000000

'key_attr.outdoor_seating'

col    cnt      prob  cume_prob
0  None  29608  0.562451  0.562451 
1  0     12945  0.245911  0.808362 
2  1     10088  0.191638  1.000000

'key_attr.parking.garage'

col    cnt      prob  cume_prob
0  0     35399  0.672461  0.672461 
1  None  14697  0.279193  0.951654 
2  1     2545   0.048346  1.000000

'key_attr.parking.lot'

col    cnt      prob  cume_prob
0  0     20769  0.394540  0.394540 
1  1     17157  0.325925  0.720465 
2  None  14715  0.279535  1.000000

'key_attr.parking.street'

col    cnt      prob  cume_prob
0  0     32443  0.616307  0.616307 
1  None  14715  0.279535  0.895842 
2  1     5483   0.104158  1.000000

'key_attr.parking.valet'

col    cnt      prob  cume_prob
0  0     36908  0.701126  0.701126 
1  None  14712  0.279478  0.980604 
2  1     1021   0.019396  1.000000

'key_attr.parking.validated'

col    cnt      prob  cume_prob
0  0     37370  0.709903  0.709903 
1  None  15118  0.287191  0.997094 
2  1     153    0.002906  1.000000

'key_attr.payment_types.amex'

col    cnt      prob  cume_prob
0  None  52628  0.999753  0.999753 
1  0     10     0.000190  0.999943 
2  1     3      0.000057  1.000000

'key_attr.payment_types.cash_only'

col    cnt      prob  cume_prob
0  None  52628  0.999753  0.999753 
1  0     11     0.000209  0.999962 
2  1     2      0.000038  1.000000

'key_attr.payment_types.discover'

col    cnt      prob  cume_prob
0  None  52628  0.999753  0.999753 
1  0     10     0.000190  0.999943 
2  1     3      0.000057  1.000000

'key_attr.payment_types.mastercard'

col    cnt      prob  cume_prob
0  None  52628  0.999753  0.999753 
1  1     10     0.000190  0.999943 
2  0     3      0.000057  1.000000

'key_attr.payment_types.visa'

col    cnt      prob  cume_prob
0  None  52628  0.999753  0.999753 
1  1     11     0.000209  0.999962 
2  0     2      0.000038  1.000000

'key_attr.price_range'

col    cnt      prob  cume_prob
0  2     21635  0.410991  0.410991 
1  1     14704  0.279326  0.690317 
2  None  10947  0.207956  0.898273 
3  3     4333   0.082312  0.980585 
4  4     1022   0.019415  1.000000

'key_attr.smoking'

col    cnt      prob  cume_prob
0  None     48553  0.922342  0.922342 
1  outdoor  1894   0.035980  0.958321 
2  no       1334   0.025341  0.983663 
3  yes      860    0.016337  1.000000

'key_attr.take-out'

col    cnt      prob  cume_prob
0  None  32341  0.614369  0.614369 
1  1     18205  0.345833  0.960202 
2  0     2095   0.039798  1.000000

'key_attr.takes_reservations'

col    cnt      prob  cume_prob
0  None  32840  0.623848  0.623848 
1  0     12658  0.240459  0.864307 
2  1     7143   0.135693  1.000000

'key_attr.waiter_service'

col    cnt      prob  cume_prob
0  None  33778  0.641667  0.641667 
1  1     11908  0.226212  0.867879 
2  0     6955   0.132121  1.000000

'key_attr.wheelchair_accessible'

col    cnt      prob  cume_prob
0  None  33114  0.629053  0.629053 
1  1     17843  0.338956  0.968010 
2  0     1684   0.031990  1.000000

'key_attr.wi-fi'

col    cnt      prob  cume_prob
0  None  34743  0.659999  0.659999 
1  no    10182  0.193423  0.853422 
2  free  7323   0.139112  0.992534 
3  paid  393    0.007466  1.000000

### Notes on above
+ most businesses accept credit cards (81%)
+ insurance is not relevant to most businesses (99.1%)
+ 99.55% of businsesses don't have age specific restrictions
    - .3% of businesses have 21+ restriction

### hours

In [13]:
%%showsql
select business_id, hours from business_raw limit 5;

business_id  \
0  "4eFIVmbeCvJGCrhzmd-VYA"   
1  "EUXT1vSFWaiD_BFmFK8rBg"   
2  "xfu2Ja977b0UA1jZZN72UA"   
3  "nuGD-Ewf7sxts2b_z_DkwQ"   
4  "eOPxvbzFTBjO5LSV3uYMug"   

                                                                                                                                                                                                                                                                                                                                               hours  
0  {"Monday": {"close": "19:00", "open": "06:00"}, "Tuesday": {"close": "19:00", "open": "06:00"}, "Friday": {"close": "19:00", "open": "06:00"}, "Wednesday": {"close": "19:00", "open": "06:00"}, "Thursday": {"close": "19:00", "open": "06:00"}, "Sunday": {"close": "14:00", "open": "07:00"}, "Saturday": {"close": "16:00", "open": "07:00"}}  
1  {}                                                                                                                                                                                                                                                                                                                                                 
2  {"Monday": {"close": "17:30", "open": "07:00"}, "Tuesday": {"close": "17:30", "open": "07:00"}, "Friday": {"close": "17:30", "open": "07:00"}, "Wednesday": {"close": "17:30", "open": "07:00"}, "Thursday": {"close": "17:30", "open": "07:00"}, "Sunday": {"close": "16:00", "open": "09:00"}, "Saturday": {"close": "17:00", "open": "07:00"}}  
3  {"Monday": {"close": "21:00", "open": "07:00"}, "Tuesday": {"close": "21:00", "open": "07:00"}, "Friday": {"close": "21:00", "open": "07:00"}, "Wednesday": {"close": "21:00", "open": "07:00"}, "Thursday": {"close": "21:00", "open": "07:00"}, "Sunday": {"close": "21:00", "open": "07:00"}, "Saturday": {"close": "21:00", "open": "07:00"}}  
4  {}

In [14]:
%%showsql
select 
    trim(both '"' from business_id) as business_id
    ,(flatten_json(hours, 'hours')).* 
from business_raw  limit 5

business_id                  key  value
0  CHYRvicu85H-gnmrZY1-Zw  hours_Monday_close   21:00
1  CHYRvicu85H-gnmrZY1-Zw  hours_Monday_open    10:00
2  CHYRvicu85H-gnmrZY1-Zw  hours_Tuesday_close  21:00
3  CHYRvicu85H-gnmrZY1-Zw  hours_Tuesday_open   10:00
4  CHYRvicu85H-gnmrZY1-Zw  hours_Friday_close   21:00

In [84]:
%%execsql
-- function to clean the resulting parsed json values
drop type if exists hours cascade;
create type hours as (
    dayofweek text
    ,ops text
);

create or replace function hours_pprint(
    business_hours text
) returns hours as
$$
    hours = business_hours.split('.')
    return [hours[1], hours[2]] # [dayofweek, open/close]
$$ language plpythonu;

In [85]:
%%execsql
drop table if exists business_hours_temp;
create temp table business_hours_temp as
select
    business_id
    ,(hours_pprint(key)).*
    ,value as hour
from (
select 
    trim(both '"' from business_id) as business_id
    ,(flatten_json(hours, 'hours')).* 
from business_raw 
) t1
distributed by (business_id);

In [86]:
%%showsql
select * from business_hours_temp limit 5;

business_id dayofweek    ops   hour
0  TLrX4FECQDFz-E45TI_9ow  Monday    close  22:00
1  TLrX4FECQDFz-E45TI_9ow  Monday    open   17:30
2  TLrX4FECQDFz-E45TI_9ow  Tuesday   close  22:00
3  TLrX4FECQDFz-E45TI_9ow  Tuesday   open   17:30
4  TLrX4FECQDFz-E45TI_9ow  Friday    close  14:30

In [87]:
%%execsql
drop table if exists business_hours;
select sugarlib.sugar($$
    create table yelp.business_hours as
    select
        business_id
        ,dayofweek
        ,sugarlib.pivot(ops,business_hours_temp,hour)
    from 
        business_hours_temp
    group by 1,2
    distributed by (business_id);
$$);

In [88]:
%%showsql
select * from business_hours limit 5;

business_id dayofweek ops_close ops_open
0  05fTEsjUPVo5jks8PpNKCQ  Friday    01:30     10:00  
1  05fTEsjUPVo5jks8PpNKCQ  Monday    20:00     10:00  
2  05fTEsjUPVo5jks8PpNKCQ  Saturday  01:30     10:00  
3  05fTEsjUPVo5jks8PpNKCQ  Sunday    22:00     08:00  
4  05fTEsjUPVo5jks8PpNKCQ  Thursday  20:00     10:00

In [77]:
%%showsql
select 
    trim(both '"' from business_id) as business_id
    ,(flatten_json(hours, 'hours')).* 
from business_raw limit 5

business_id                  key  value
0  Y68Ysab60gJhX7YBjESgBg  hours.Tuesday.close  17:00
1  Y68Ysab60gJhX7YBjESgBg  hours.Tuesday.open   08:00
2  Y68Ysab60gJhX7YBjESgBg  hours.Friday.close   17:00
3  Y68Ysab60gJhX7YBjESgBg  hours.Friday.open    08:00
4  Y68Ysab60gJhX7YBjESgBg  hours.Monday.close   17:00

In [89]:
%%execsql
drop table if exists business_hours_temp;
create temp table business_hours_temp as
select 
    trim(both '"' from business_id) as business_id
    ,(flatten_json(hours, 'hours')).* 
from business_raw 
distributed by (business_id);

In [90]:
%%execsql
drop table if exists business_hours_wide;
select sugarlib.sugar($$
    create table yelp.business_hours_wide as
    select
        business_id
        ,sugarlib.pivot(key,business_hours_temp,value)
    from 
        business_hours_temp
    group by 1
    distributed by (business_id);
$$);

In [91]:
%%showsql
select * from business_hours_wide limit 5;

business_id key_hours.friday.close key_hours.friday.open  \
0  08xNB6lB-2kMrj5bqJaEMA  00:00                  10:00                  
1  0btslkqA-YJQjLH6GFjU-g  21:00                  09:00                  
2  0gABc2O0CsMqhT35LPHfNQ  18:00                  09:00                  
3  0m-dEwpqVNO4mgZ1mM7Wew  17:00                  08:00                  
4  0NeXPGy3_N9T4IeHde7iGQ  22:00                  11:00                  

  key_hours.monday.close key_hours.monday.open key_hours.saturday.close  \
0  00:00                  10:00                 00:00                     
1  21:00                  09:00                 21:00                     
2  18:00                  09:00                 18:00                     
3  17:00                  08:00                 17:00                     
4  21:00                  11:00                 22:00                     

  key_hours.saturday.open key_hours.sunday.close key_hours.sunday.open  \
0  10:00                   00:00                  10:00                  
1  09:00                   18:00                  09:00                  
2  09:00                   16:00                  10:00                  
3  08:00                   None                   None                   
4  11:00                   20:00                  11:00                  

  key_hours.thursday.close key_hours.thursday.open key_hours.tuesday.close  \
0  00:00                    10:00                   00:00                    
1  21:00                    09:00                   21:00                    
2  18:00                    09:00                   18:00                    
3  17:00                    08:00                   17:00                    
4  21:00                    11:00                   21:00                    

  key_hours.tuesday.open key_hours.wednesday.close key_hours.wednesday.open  
0  10:00                  00:00                     10:00                    
1  09:00                  21:00                     09:00                    
2  09:00                  18:00                     09:00                    
3  08:00                  17:00                     08:00                    
4  11:00                  21:00                     11:00

### neighborhoods

In [52]:
%%showsql
select business_id, neighborhoods from business_raw where neighborhoods ilike '%,%' limit 5;

business_id                 neighborhoods
0  "Pc932k0ZDQYxjt3RflTx8w"  ["Southeast", "University"] 
1  "Dhz3Qj3RLOT1ublaAqBw9g"  ["Tollcross", "The Meadows"]
2  "TMhHqj5KnkxOBR9vdM84qg"  ["Cotswold", "Eastland"]    
3  "rrD6XzX6fQ-aBnGczJIn-Q"  ["Old Town", "Royal Mile"]  
4  "I0U4MQTiSNdGmpYoNcrimg"  ["West End", "New Town"]

In [32]:
%%execsql
drop table if exists business_neighborhoods;
create table business_neighborhoods as
select trim(both '"' from business_id) as business_id, trim(both ' ' from split_array(neighborhoods)) as neighborhood
from business_raw 
distributed by (business_id);

In [24]:
%%showsql
select * from business_neighborhoods limit 5;

business_id neighborhood
0  0ilWo2kWlmP_Nv-P2rcNOA  Eken Park  
1  nUISXnW2Jo9ovnoFBIUY_Q  Bloomfield 
2  DmRw9flaIQ1VTdmjGspN6A  Westside   
3  T4D7wX6GuCYChwRmser-0A  Brookline  
4  NMqAOyNp6J3rH5B_RZBbbw  Leith

### Create meta data table for restaurants

In [34]:
%%execsql
drop table if exists businesses;
create table businesses as
with business as (
    select 
        trim(both '"' from business_id) as business_id
        ,trim(both '"' from city) as city
        ,trim(both '"' from full_address) as full_address
        ,latitude::double precision
        ,longitude::double precision
        ,trim(both '"' from name) as name
        ,cast_array(categories) as categories
        ,case when open = 'true' then 1 else 0 end as open
        ,review_count::integer
        ,stars::double precision
        ,cast_array(neighborhoods) as neighborhoods
        ,trim(both '"' from state) as state
    from 
        business_raw as business
)
select
    *
from 
    business 
left join
    business_hours_wide hours
using (business_id)
left join 
    business_attributes attr
using (business_id)
distributed by (business_id);

In [35]:
%%showsql
select * from businesses limit 5;

business_id       city  \
0  P7wKlbSRsoRrfH3DFOY08A  Phoenix     
1  1MApscjpyygvanCdaesG3g  Charlotte   
2  YHDKMl5ij09wncT3eTX0yg  Las Vegas   
3  tL8QOD03ibKRf5jeEuMTQQ  Chandler    
4  xx4Qz0ILrTzNBBdVO1IS5Q  Las Vegas   

                                                           full_address  \
0  2508 N Central Ave\nPhoenix, AZ 85004                                  
1  4718 Sharon Rd.\nSharon Corners S/C\nSouth Park\nCharlotte, NC 28210   
2  3113 N Rainbow Blvd\nLas Vegas, NV 89108                               
3  1000 N Arizona Ave\nChandler, AZ 85225                                 
4  530 Las Vegas Blvd S\nDowntown\nLas Vegas, NV 89101                    

    latitude   longitude                             name  \
0  33.475606 -112.074002  International House of Pancakes   
1  35.147245 -80.832369   Men's Wearhouse                   
2  36.216485 -115.242364  Famous Nails                      
3  33.320904 -111.842059  Pizza Hut                         
4  36.163880 -115.144546  Wells Fargo Bank                  

                                         categories  open  review_count  \
0  [Breakfast & Brunch, Restaurants]                 1     43             
1  [Men's Clothing, Fashion, Shopping, Formal Wear]  1     6              
2  [Beauty & Spas, Nail Salons]                      1     7              
3  [Italian, Pizza, Chicken Wings, Restaurants]      1     11             
4  [Banks & Credit Unions, Financial Services]       0     3              

   stars neighborhoods state key_hours.friday.close key_hours.friday.open  \
0  3.0    None          AZ    None                   None                   
1  3.0    [South Park]  NC    21:00                  10:00                  
2  2.0    None          NV    None                   None                   
3  2.0    None          AZ    None                   None                   
4  4.0    [Downtown]    NV    None                   None                   

  key_hours.monday.close key_hours.monday.open key_hours.saturday.close  \
0  None                   None                  None                      
1  21:00                  10:00                 21:00                     
2  None                   None                  None                      
3  None                   None                  None                      
4  None                   None                  None                      

  key_hours.saturday.open key_hours.sunday.close key_hours.sunday.open  \
0  None                    None                   None                   
1  09:30                   18:00                  11:00                  
2  None                    None                   None                   
3  None                    None                   None                   
4  None                    None                   None                   

  key_hours.thursday.close key_hours.thursday.open key_hours.tuesday.close  \
0  None                     None                    None                     
1  21:00                    10:00                   21:00                    
2  None                     None                    None                     
3  None                     None                    None                     
4  None                     None                    None                     

  key_hours.tuesday.open key_hours.wednesday.close key_hours.wednesday.open  \
0  None                   None                      None                      
1  10:00                  21:00                     10:00                     
2  None                   None                      None                      
3  None                   None                      None                      
4  None                   None                      None                      

  payment_types.visa hair_types_specialized_in.kids  \
0  None               None                            
1  None               None                            
2  None               None          

## Review Data

In [4]:
%%showsql
select * from yelp.review_raw limit 5;

business_id   review_date                 review_id stars  \
0  "LHyGQnuWR2q44bsrw1ZLAg"  "2009-03-10"  "0vOc47f2dyOdRdJlhuQ-wA"  4      
1  "LHyGQnuWR2q44bsrw1ZLAg"  "2010-02-25"  "loWTkwyzwf-yvr6-6-w7vQ"  3      
2  "LHyGQnuWR2q44bsrw1ZLAg"  "2010-03-19"  "9JnE3t2o3ILxMxgAfijnxg"  2      
3  "LHyGQnuWR2q44bsrw1ZLAg"  "2011-03-20"  "iII3bV909_MlfFIUYkDudQ"  3      
4  "LHyGQnuWR2q44bsrw1ZLAg"  "2011-06-24"  "Yj6zM-3fUQgKRRWWUMCKcA"  4      

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           review_text  \
0  "Lots of options. The sandwiches are moderately priced and good. Avoid the cupcakes - they are dry."                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
1  "Alright, bad news first. Sadly, the Bagel Factory wasn't quite as good as La Bagel Delight (my favorite Brooklyn bagel shop/deli). I went in hoping for something that would rival LBD's Mr. Bagel, and I'm sorry to say that in that regard, the Bagel Factory fell a bit short.\n\nOn the other hand, this is Pittsburgh, not New York City, and the Bagel Factory is by far the best bagel I've had here. Brueggers and Einstein don't even compare.\n\nThe egg in my bacon, egg and cheese bagel sandwich left a lot to be desired. It definitely seemed like a premade egg product that was simply heated for my sandwich. Fresh, scrambled eggs would have earned the Bagel Factory 4 stars here. I imagine this place would be excellent for the standard bagel and cream cheese.\n\nThe service was friendly and fast, the place was clean."                                                                                                                                                                                 
2  "It is rather unfortunate that Pittsburgh doesn't have ANY privately owned strictly bagel joints. Bagels, in my opinion, are one of the best breakfast alternatives to actually sitting down to a nice breakfast. I want this - being a student, basically full-time server, volunteering - my life pretty much requires a bagel breakfast every now and then. However, I'm in Pittsburgh, where can I go?\nEisteins? Brueggers? These place are virtually the same. Panera? I used to work here so not only do I dislike the taste, but I have a natural aversi

### votes

In [5]:
%%execsql
drop table if exists review_votes_temp;
create temp table review_votes_temp as
select 
review_id 
,business_id 
,(flatten_json(votes, 'votes')).* 
from review_raw 
distributed by (review_id);

In [6]:
%%showsql
select * from review_votes_temp limit 5

review_id               business_id           key value
0  "ISphyYy9XO_5-CY3M228GA"  "e5J3s-xxs1S6PPHJhldBEA"  votes_funny   0   
1  "ISphyYy9XO_5-CY3M228GA"  "e5J3s-xxs1S6PPHJhldBEA"  votes_useful  0   
2  "ISphyYy9XO_5-CY3M228GA"  "e5J3s-xxs1S6PPHJhldBEA"  votes_cool    0   
3  "_uJeFDqNqcgzUeKLNDJliw"  "Og03_vbBCVAFNHXgdflRLw"  votes_funny   0   
4  "_uJeFDqNqcgzUeKLNDJliw"  "Og03_vbBCVAFNHXgdflRLw"  votes_useful  0

In [7]:
%%execsql
drop table if exists review_votes;
select sugarlib.sugar($$
    create table review_votes as
    select review_id, business_id
        ,sugarlib.pivot(key,review_votes_temp,value)
    from 
        review_votes_temp
    group by review_id, business_id
    distributed by (review_id);
$$);

In [9]:
%%showsql
select * from review_votes limit 5

review_id               business_id key_votes_cool  \
0  "0035PeaN1SlPq4JiEGU_dg"  "l2lpBXx8jKl87J2szyJRuQ"  1               
1  "00DkGNkWv6uTieRIVJ-opg"  "OedV_QX8aD7MeLqshhWkCQ"  0               
2  "00fhWT2o1qP3lUN1nxbm0w"  "QkAv4k5aB6SwNRNK7q19ug"  0               
3  "00tBGNrc_pQHFaJx4vdL_w"  "8KSK12p77qHb4gvN6DAncg"  1               
4  "01B60AUs02ckmliQqRBjPg"  "eFIwRnbZpToIwWOYWWJ7nA"  0               

  key_votes_funny key_votes_useful  
0  0               1                
1  0               0                
2  0               0                
3  0               2                
4  0               0

### Cleaned review data

In [4]:
%%execsql
drop table if exists review;
create table review as
select 
    trim(both '"' from t1.business_id) as business_id
    ,trim(both '"' from t1.review_id) as review_id 
    ,trim(both '"' from review_date)::date as review_date
    ,stars::integer
    ,trim(both '"' from review_text) as review_text 
    ,trim(both '"' from user_id) as user_id
    ,key_votes_cool::int as votes_cool
    ,key_votes_funny::int as votes_funny
    ,key_votes_useful::int as votes_useful 
from 
    review_raw t1
left join
    review_votes t2
using (review_id, business_id)
distributed by (review_id);

In [53]:
%%showsql
select * from review limit 3;

business_id               review_id review_date  stars  \
0  EEuGTah-oEssSl3o6pK-ZQ  wVLvNJHpmlrZ-m8jnh-cGQ  2008-10-07  3       
1  9NL2GL4jSPVj9SKijH5PqA  l9CVLRgt5uzQdakPQMffOg  2014-11-18  4       
2  105yyfGnNxJPPFbdIgBHLw  oUo17eaxek-rVwtnNP20LQ  2007-08-20  5       

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                review_text  \
0  The pros: Dinosaurs in Their Time is pretty cool.  The years of renovation were worth it - the former dinosaur hall is completely transformed.  Remember how everything was in straight lines and the dinosaurs were up on their big pedestals?  Now the exhibit is sort of S-shaped, the walls are painted with colorful murals, the support columns are disguised as trees, there are plants around the dinosaurs and dead leaves underneath.  Underneath every dinosaur is a nameplate with a lot of information and a touchscreen with even MORE information, including a section on how the Carnegie Museum researchers contributed to the discovery or naming of this dinosaur.  There is a lot more natural light than I remember, and there are 2 balconies that overlook the exhibits for more photo opportunities.  There is also an extensive timeline of the Carnegie Museum's expeditions and contributions to paleontology, and a timelapse video of the renovations.\n\nThe cons: The rest of the history museum is pretty lame compared to this new and awesome exhibit.  The biggest examples are the 2nd floor Halls of African and North American Wildlife and the 3rd floor Birds hall.  Big cases of taxidermy in a wood-paneled room with dim lighting - hardly thrilling.  The 3rd floor Polar World is starting to look dated too.  Even the fonts used date the exhibits.\n\nAdmission is now $15 for adults, $11 for students (free for Pitt students!).  This does get you into the Carnegie Museum of Art also, but I'm usually tired after doing the whole history museum so I never get over to the art side.  I would recommend seeing the dinosaurs, then skipping the rest of the history museum if you've been there before.  The Dinosaurs in Their Time would get five stars on its own, but the rest of the museum is lagging behind unfortunately.   
1  I haven't been to Kuhn's in months and forgot how much I prefer it over the 

## User Data

In [5]:
%%showsql
select * from yelp.user_raw limit 5;

average_stars  \
0  4.37           
1  3.11           
2  3.0            
3  3.49           
4  4.17           

                                                                                                                                compliments  \
0  {"funny": 4, "plain": 6, "writer": 1, "note": 1, "hot": 1, "cool": 5}                                                                      
1  {"writer": 1}                                                                                                                              
2  {}                                                                                                                                         
3  {"profile": 8, "funny": 22, "cute": 9, "plain": 70, "writer": 41, "list": 4, "note": 27, "photos": 50, "hot": 86, "cool": 90, "more": 8}   
4  {}                                                                                                                                         

          elite fans  \
0  []            1     
1  []            0     
2  []            0     
3  [2009, 2010]  32    
4  []            0     

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   

### compliments

In [62]:
%%execsql
drop table if exists user_compliments_temp;
create temp table user_compliments_temp as
select 
    trim(both '"' from user_id) as user_id
    ,(flatten_json(compliments, 'compliments')).*
from 
    user_raw
distributed by (user_id);

In [63]:
%%showsql
select * from user_compliments_temp limit 5;

user_id                 key value
0  09XpyD1prjqKq4nS3XkKpA  compliments.plain   2   
1  iEoPIDFBxiEuCVG8GzyENg  compliments.funny   4   
2  iEoPIDFBxiEuCVG8GzyENg  compliments.plain   1   
3  iEoPIDFBxiEuCVG8GzyENg  compliments.writer  3   
4  iEoPIDFBxiEuCVG8GzyENg  compliments.note    4

In [16]:
%%showsql
select key, count(*) from user_compliments_temp group by 1 order by 2 desc;

key  count
0   compliments_plain    84007
1   compliments_cool     77966
2   compliments_note     73508
3   compliments_writer   59562
4   compliments_hot      55896
..              ...        ...
6   compliments_more     44919
7   compliments_photos   29017
8   compliments_cute     22080
9   compliments_profile  19857
10  compliments_list     11556

[11 rows x 2 columns]

In [64]:
%%execsql
drop table if exists user_compliments;
select sugarlib.sugar($$
    create table yelp.user_compliments as
    select user_id as user_id
        ,sugarlib.pivot(key,user_compliments_temp,value)
    from 
        user_compliments_temp
    group by user_id
    distributed by (user_id);
$$);

In [65]:
%%showsql
select * from user_compliments limit 5;

user_id key_compliments.cool key_compliments.cute  \
0  01nTwhHBaw8AHjQDb7JGBw  None                 None                  
1  05qgjGUgH9Z8r1XRDRpkww  1                    None                  
2  06YTa8A3QhWXv4l6Htr91A  4                    None                  
3  07Ej2jE4VRoT0EmGtz8P5Q  None                 None                  
4  07-MOraFwL_8CRJuzVYagA  1                    None                  

  key_compliments.funny key_compliments.hot key_compliments.list  \
0  None                  1                   None                  
1  None                  None                None                  
2  5                     None                None                  
3  1                     None                None                  
4  1                     None                None                  

  key_compliments.more key_compliments.note key_compliments.photos  \
0  None                 None                 None                    
1  None                 None                 None                    
2  1                    9                    None                    
3  None                 None                 None                    
4  1                    1                    None                    

  key_compliments.plain key_compliments.profile key_compliments.writer  
0  3                     None                    None                   
1  None                  None                    None                   
2  5                     None                    1                      
3  None                  None                    None                   
4  None                  None                    1

### elite

In [70]:
%%showsql
select * from user_raw limit 5;

average_stars  \
0  4.0            
1  3.54           
2  3.36           
3  5.0            
4  4.23           

                                                                                          compliments  \
0  {"note": 1}                                                                                          
1  {"note": 1, "plain": 5, "writer": 1, "cool": 1}                                                      
2  {}                                                                                                   
3  {}                                                                                                   
4  {"funny": 14, "plain": 18, "writer": 29, "note": 6, "photos": 3, "hot": 63, "more": 4, "cool": 52}   

  elite fans  \
0  []    0     
1  []    4     
2  []    0     
3  []    0     
4  []    5     

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      friends  \
0  ["4McyrxqtF0U61vijemLWHg", "4eM4nGh58BsLlm5lUXvuiQ"]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
1  ["7b134wSLX8gN7G2sbLPZoQ", "hc4D2WkjKG8mk1C9e0fsQQ", "UoaC8YC4Z_rssccxV8nI-g", "F7VViCY5ja6RVykGb2P6qA", "CWzxV9y-UtHdnJASkFstzA"]                                                

In [33]:
%%execsql
create or replace function cast_array(
    arr text
) returns text[] as 
$$
    import ast
    result = ast.literal_eval(arr)
    if result:
        return [i.strip() for i in result]
    else:
        return None
$$ language plpythonu;

In [23]:
%%showsql
select cast_array(elite) as elite from user_raw limit 5;

elite
0  None
1  None
2  None
3  None
4  None

In [296]:
%%execsql
drop table if exists user_elite;
create table user_elite as
select 
    trim(both '"' from user_id) as user_id
    ,split_array(elite) as elite
from 
    user_raw
distributed by (user_id);

In [69]:
%%showsql
select * from user_elite limit 5

user_id  elite
0  CdrOG7bBwAgHfkskPlA8xA  2011 
1  CdrOG7bBwAgHfkskPlA8xA   2012
2  CdrOG7bBwAgHfkskPlA8xA   2013
3  CdrOG7bBwAgHfkskPlA8xA   2014
4  CdrOG7bBwAgHfkskPlA8xA   2015

### friends

In [299]:
%%execsql
create table user_friends as
select 
    trim(both '"' from user_id) as user_id
    ,split_array(friends) as friends
from 
    user_raw
distributed by (user_id);

In [300]:
%%showsql
select * from user_friends limit 5;

user_id                  friends
0  MjIlb4iQMEpEEDGNKoo97w  LGwoVXv6RwB7tEjgyeOZWw 
1  qKfceRHpoYNIDV1HryX_hw  vWnUf5VXx9S7cbgomxsOww 
2  GJ45sYsFmWuou9TZoTrJJg  VzAHWXVOIqylfCtIRXwU9w 
3  GJ45sYsFmWuou9TZoTrJJg   7x3-B0yaO_hqXm42VKStzw
4  GJ45sYsFmWuou9TZoTrJJg   1ku6epbYu8IN2hmleuHvQg

### votes

In [56]:
%%execsql
drop table if exists user_votes_temp;
create temp table user_votes_temp as 
select
    user_id
    ,(votes).key as key
    ,(votes).value::integer as value
from (
    select
        trim(both '"' from user_id) as user_id
        ,(flatten_json(votes, 'votes')) as votes
    from 
        user_raw
) t1
distributed by (user_id);

In [57]:
%%showsql
select * from user_votes_temp limit 5;

user_id           key  value
0  atvdPb748A_vq1uU6e6z4Q  votes.funny   19   
1  atvdPb748A_vq1uU6e6z4Q  votes.useful  58   
2  atvdPb748A_vq1uU6e6z4Q  votes.cool    23   
3  edRbqZjCFLbOdMGDtnl5dQ  votes.funny   1    
4  edRbqZjCFLbOdMGDtnl5dQ  votes.useful  6

In [58]:
%%execsql
drop table if exists user_votes;
select sugarlib.sugar($$
    create table yelp.user_votes as
    select user_id as user_id
        ,sugarlib.pivot(key,user_votes_temp,value)
    from 
        user_votes_temp
    group by user_id
    distributed by (user_id);
$$);

In [59]:
%%showsql
select * from user_votes limit 5;

user_id  key_votes.cool  key_votes.funny  key_votes.useful
0  01jockG1cl78WWfcripIiQ  1               1                1               
1  02-khcM_JJOHGU3Dk5EQ5A  0               0                2               
2  04CG8ITB9L-7p1GNUxOoew  0               0                2               
3  04Vn_yTfjddj3pBW6yF4qA  10              7                12              
4  04WAMdJ1ddLLO1fG-EnA4A  14              9                44

In [73]:
%%execsql
drop table if exists users;
create table users as 
with userdata as (
select 
    average_stars::double precision
    ,fans::integer
    ,trim(both '"' from name) as name 
    ,review_count::integer 
    ,trim(both '"' from yelping_since) as yelping_since
    ,trim(both '"' from user_id) as user_id
    ,cast_array(elite) as elite
    ,cast_array(friends) as friends
from 
    user_raw 
)
select 
    *
from 
    userdata
left join 
    user_compliments usercomps
using (user_id)
left join user_votes uservotes
using (user_id)
distributed by (user_id);

In [121]:
%%showsql
select * from users limit 5

user_id  average_stars  fans     name  review_count  \
0  gywsvSBCL6yCOmuXhluuBw  4.31           0     C        17             
1  LyyualA8h35oLeOUxJxFbg  3.40           1     Sarah    65             
2  2h_EEqVKokJ1yqU8WAtXhA  3.60           0     Twinkle  20             
3  YKyJitb6wgvX2GAIyOcJ3g  4.00           0     Alex     1              
4  Ip123mRss7vVTngqearVXA  3.50           0     Nikita   53             

  yelping_since elite  \
0  2007-02       []     
1  2009-02       []     
2  2009-01       []     
3  2012-11       []     
4  2011-07       []     

                                                                    friends  \
0  [pmPV9jgDyad6DK-jxZQ0PA, dw1zh9oNtZSSdhn9lrVdvA]                           
1  []                                                                         
2  []                                                                         
3  []                                                                         
4  [OqCgbvz3ZVsXzQreFvIWmw, clCE4R0bAjwp_L4IX3BvUg, meq4JCIvdj-zYbRS0s2ODQ]   

  key_compliments.cool key_compliments.cute key_compliments.funny  \
0  1                    None                 None                   
1  None                 None                 None                   
2  None                 None                 None                   
3  None                 None                 None                   
4  None                 None                 None                   

  key_compliments.hot key_compliments.list key_compliments.more  \
0  1                   None                 None                  
1  None                None                 1                     
2  None                None                 None                  
3  None                None                 None                  
4  None                None                 None                  

  key_compliments.note key_compliments.photos key_compliments.plain  \
0  None                 None                   None                   
1  None                 None                   None                   
2  None                 None                   None                   
3  None                 None                   None                   
4  None                 None                   None                   

  key_compliments.profile key_compliments.writer  key_votes.cool  \
0  None                    None                   8                
1  None                    None                   9                
2  None                    None                   7                
3  None                    None                   0                
4  None                    None                   3                

   key_votes.funny  key_votes.useful  
0  2                42                
1  8                17                
2  3                7                 
3  0                0                 
4  4                22

## Check In Data

In [6]:
%%showsql
select * from yelp.checkin_raw limit 5;

business_id  \
0  "ra0evfke9lyf5sWzvKUbWQ"   
1  "iWPMRDI_MZDZwx0GCJ_imA"   
2  "HB8EjFxgesxahG4rvpGzIA"   
3  "VTMqRsk1dXJJaReAQLKIsg"   
4  "1iPwf-E4e-hA3Td1QonzLQ"   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                             checkin_info  \
0  {"10-4": 1, "15-5": 1, "10-2": 1, "12-3": 1, "18-1": 1, "18-2": 1, "18-4": 1, "18-5": 1, "19-4": 1, "19-2": 1, "11-5": 1, "11-4": 1, "11-1": 1, "13-6": 1, "11-2": 1, "9-5": 2, "6-4": 2, "12-5": 1, "6-2": 4, "6-3": 1, "6-0": 1, "13-3": 1, "9-3": 1, "5-1": 1, "9-4": 1, "13-2": 1, "7-3": 1, "7-2": 1, "7-1": 2, "9-6": 2, "8-4": 1, "7-5": 1, "17-6": 1, "16-0": 2, "16-1": 1, "16-5": 1, "10-0": 1, "10-1": 1, "8-6": 1, "10-3": 1, "8-0": 1, "10-5": 3, "10-6": 1, "11-0": 1}     
1  {"19-5": 1, "19-1": 1, "15-5": 1, "13-0": 1, "17-3": 1, "12-3": 1}                                                                                                                                                                                                                                                                                                                                                                                                                       
2  {"16-4": 1, "13-2": 1, "17-5": 1, "17-4": 1, "17-2": 1, "13-5": 1, "10-5": 2}                                                                                                                                                                                                                                                                                                                                                                                                            
3  {"12-5": 1, "11-4": 1, "17-0": 1, "19-1": 1, "11-5": 1, "17-6": 1, "17-3": 1, "17-2": 1, "13-4": 2, "16-5": 1, "21-0": 1, "22-2": 1, "13-2": 1, "12-1": 2}                                                                                                                                                                                                                                                                                                                               
4  {"13-5": 7, "10-5": 4, "11-2": 2, "15-2": 1, "15-5": 1, "15-4": 1, "10-0": 1, "18-1": 2, "18-2": 1, "18-3": 2, "18-5": 1, "14-4": 2, "14-5": 2, "14-6": 1, "14-1": 1, "14-3": 1, "17-1": 1, "19-1": 1, "19-3": 1, "19-2": 1, "11-5": 1, "11-4": 1, "13-1": 1, "11-1": 2, "11-3": 1, "13-4": 2, "12-5": 3, "12-2": 3, "12-3": 1, "22-2": 1, "9-5": 1, "20-1": 3, "20-2": 1, "8-1": 1, "16-3": 1, "16-1": 2, "17-3": 2, "17-2": 4, "16-4": 1, "8-4": 2, "10-1": 1, "10-2": 2, "21-1": 1}   

        type  
0  "checkin"  
1  "checkin"  
2  "checkin"  
3  "checkin"  
4  "checkin"

### checkin_info

In [124]:
%%execsql
drop type if exists checkin cascade;
create type checkin as (
    hour integer 
    ,day text
);

create or replace function checkin_pprint(
    checkin text
) returns checkin as 
$$
    import calendar as cal
    val = checkin.split('.')[1].split('-')
    return [int(val[0]), cal.day_name[int(val[1])]]

$$ language plpythonu;

In [132]:
%%execsql
drop table if exists checkins;
create table checkins as
select 
    business_id
    ,(checkin_pprint(key)).*
    ,value::int as num_checkins
from (
select 
    trim(both '"' from business_id) as business_id
    ,(flatten_json(checkin_info, 'checkin')).*
from 
    checkin_raw
) t1
distributed by (business_id);

In [133]:
%%showsql
select * from checkins limit 5

business_id  hour        day  num_checkins
0  RzhkBV3z_RWw9WHcidKQlw  10    Monday     2           
1  RzhkBV3z_RWw9WHcidKQlw  10    Saturday   1           
2  RzhkBV3z_RWw9WHcidKQlw  23    Tuesday    3           
3  RzhkBV3z_RWw9WHcidKQlw  13    Friday     2           
4  RzhkBV3z_RWw9WHcidKQlw  17    Wednesday  1

## Tip Data

In [7]:
%%showsql
select * from yelp.tip_raw limit 5;

business_id      tip_date likes  \
0  "YwUTwzYD7G25DIICSVqFBQ"  "2013-01-11"  0      
1  "YwUTwzYD7G25DIICSVqFBQ"  "2014-11-01"  0      
2  "YwUTwzYD7G25DIICSVqFBQ"  "2010-07-09"  0      
3  "RfR5yc4j9coIPBuxRsOW8Q"  "2010-06-26"  0      
4  "RfR5yc4j9coIPBuxRsOW8Q"  "2010-07-23"  0      

                                                                                                  tip_text  \
0  "It's in China town and the customer didn't fall in live with these girls so he came back to his room."   
1  "Video poker, topless dancers...enough said"                                                              
2  "Well, you get what you pay for..."                                                                       
3  "Again...."                                                                                               
4  "Get the Adobo Fried Rice Omelette! Won't disappoint! Raimen soup (extra hot) not bad!"                   

    type                   user_id  
0  "tip"  "DEgml5fKy8H3GgfI7k0WGA"  
1  "tip"  "qQuMyLtoSwiKNxp57h_sDg"  
2  "tip"  "-m8iD48kKeEF2nmrGGaoUQ"  
3  "tip"  "8DdcyZR-THOgH60bIyEnfg"  
4  "tip"  "rttCl3o0bLAbFG60Jmv2UA"

In [134]:
%%execsql
drop table if exists tips;
create table tips as 
select
    trim(both '"' from business_id) as business_id
    ,(trim (both '"' from tip_date))::date as tip_date
    ,likes::integer
    ,trim(both '"' from tip_text) as tip_text
    ,trim(both '"' from user_id) as user_id
from 
    tip_raw
distributed by (business_id);

In [135]:
%%showsql
select * from tips limit 5;

business_id    tip_date  likes  \
0  Zx8_4zKdDBSO3qGrkukBIA  2009-11-08  0       
1  Zx8_4zKdDBSO3qGrkukBIA  2009-12-28  0       
2  Zx8_4zKdDBSO3qGrkukBIA  2010-01-24  0       
3  Zx8_4zKdDBSO3qGrkukBIA  2010-01-28  0       
4  Zx8_4zKdDBSO3qGrkukBIA  2010-02-02  0       

                                                                            tip_text  \
0  12Q FTW!                                                                            
1  Avoid the lunch crowds / lines by coming before 11:30 am.                           
2  Jade red chicken!!                                                                  
3  Jade red!                                                                           
4  Suicide plate time #13 9 jfr and black beans.  You owe it to yourself to try this   

                  user_id  
0  PNAkqna65eJq8Od5dl1iDA  
1  THC7I7imU3jkhYN1ibSH3w  
2  yL9YKQivaqxQClwMxhPajg  
3  sALB8-F04S9VcZMF_GkGUA  
4  N3L6SDymSRkYQtTiSDmXSg

## Final Parsed Tables

### Business Data
+ businesses
+ business_attributes
+ business_categories
+ business_hours
+ business_neighborhoods


### Review Data
+ review

### User Data
+ users
+ user_elite
+ user_friends

### Checkin Data
+ checkin_info

### Tip Data
+ tip